lib/spack/spack/cmd/__init__.py

```diff
@@ -80,46 +84,54 @@ def remove_options(parser, *options):
                 break
 
 
-def get_module(cmd_name):
-    """Imports the module for a particular command name and returns it.
+def get_module_from(cmd_name, namespace):
+    """Imports the module for a particular command from the specified namespace.
 
     Args:
         cmd_name (str): name of the command for which to get a module
             (contains ``-``, not ``_``).
+        namespace (str): namespace for command.
     """
     pname = python_name(cmd_name)
-
-    try:
-        # Try to import the command from the built-in directory
-        module_name = "%s.%s" % (__name__, pname)
-        module = __import__(module_name,
-                            fromlist=[pname, SETUP_PARSER, DESCRIPTION],
-                            level=0)
-        tty.debug('Imported {0} from built-in commands'.format(pname))
-    except ImportError:
-        module = spack.extensions.get_module(cmd_name)
-        if not module:
-            raise
+    module_name = '{0}.cmd.{1}'.format(namespace, pname)
+    module = __import__(module_name,
```

Comment:
It may be worth catching this and testing to see if the failure is because the user didn't add `cmd` to the directory structure.

Reply:
I'm not sure that I think that would be the best thing to do, here. The `__import__()` call could raise an `ImportError` for a number of reasons, of which the omission of `cmd` from the directory structure is only one possibility. However, think that the `ImportError` will show the full specification of the module that failed to be imported, and this will include the expected `cmd` component. During development of a command extension the import is likely to fail for this reason only once, and due to Python errors in the module code multiple times. Catching the exception and then manually looking for all the places the command *might* be (`path/X.py` _vs_ `path/cmd/X.py` for all paths in the Python search path) just in case this is the problem seems like going a long way for little extra benefit, but I'm open to being convinced otherwise.

Reply:
My main worry is that a user would commonly forget to include `cmd` as part of the directory structure and would spend some time being confused by the `ImportError`, am I missing somewhere that the directory structure is verified before this?

Reply:
It's not verified, but I'm having trouble imagining any confusion, as the error message should include the search namespace, viz:
```console
ImportError: No module named bad.cmd.hello
```
The only place I can imagine putting any reliable structure verification is checking that the argument to `find_commands()` ends in a `cmd` path component, but that wouldn't catch someone adding subcommands by hand.

Reply:
IMO, it's far more likely that the user will forget to put `__init__.py` files in all the right places for the import to work with namespaces (ask me how I know, after having wasted >1h trying to work out why a test failed). If we're given a path, we could insert it at the front of `sys.path` and ensure `___init__.py` files all the way through to `cmd/`, but absent that there's no way to be general.

Reply:
> as the error message should include the search namespace, viz:
> `ImportError: No module named bad.cmd.hello`

I'm thinking this would be a problem. I think you're assuming a user would analyze every part of that message equally and then focus on `cmd`. I think they would see `ImportError` and get frustrated.

> IMO, it's far more likely that the user will forget to put __init__.py files in all the right places for the import to work with namespaces (ask me how I know, after having wasted >1h trying to work out why a test failed). If we're given a path, we could insert it at the front of sys.path and ensure ___init__.py files all the way through to cmd/, but absent that there's no way to be general.

This is a good point and also possibly worth considering, but I'm trying to think whether verifying the import system by checking the file system is a sensible practice. Overall I'm inclined to say the `cmd` issue worries me more because it is specific to Spack and its handling of custom command scopes.

Reply:
If we're going to delegate to the import system (as, I would think, is sensible), then we shouldn't be relying on having a path specified, and a filesystem check of every component  in `sys.path` seems a little heavyweight to me. I'm currently rewriting `get_module_from()` to take an optional path and verify `cmd` and `__init__.py` presence if specified there and also in `find_commands()`. Beyond that though, I'm having trouble seeing what we can usefully do. Do you have any suggestions? Am I missing or misunderstanding something about the problem and likely solutions?

Reply:
@scheibel @chissg 

> My main worry is that a user would commonly forget to include cmd as part of the directory structure and would spend some time being confused by the ImportError, am I missing somewhere that the directory structure is verified before this?

Wouldn't this be better served by a Spack command that helps managing and developing extensions? What I mean is something like (*names are just for illustrative purposes*):
```console
$ spack extensions audit <path-of-an-extension> # Reports errors and warnings on the extension
$ spack extensions get <url> # Downloads the extension and registers it automatically
$ spack extensions remove <name> # Remove a registered extension
```
If we have something like that we could catch a broader exception like `ImportError` and suggest people to run an audit on their extension to diagnose the problem. Does that seem a sensible approach?

Reply:
@alazo Yes, the issue mentioned is related to `get_module_from()`. Note that we are talking about _sub-commands_ here, and sub-commands implemented in different modules, at that. I'm also not necessarily confining the idea to sub-commands of extensions, either: `spack module` would certainly be a candidate for this treatment. Please see below for more discussion.

Reply:
> Yes, the issue mentioned is related to get_module_from(). Note that we are talking about sub-commands here, and sub-commands implemented in different modules, at that.

Sorry, I might not be understanding what you mean here. Sub-commands are *already* possible as far as I can tell. That was the primary purpose of #11209. There's a test:

https://github.com/spack/spack/blob/b4862be875dee58187a670e08c4d79f2b49f8cb9/lib/spack/spack/test/cmd_extensions.py#L105-L111

That constructs sub-commands and makes sure you can invoke:

- `spack hello world`
- `spack hello folks`
- `spack hello global`

and checks the expected output. Is there something still missing?

> Please see below for more discussion

I'm also not finding any discussion below (but it might be that I'm too tired :sleeping: ) Can you post a link?

Reply:
@alalazo the PR description has been updated and may be more informative (https://github.com/spack/spack/pull/11145#issue-268805693): one of the primary additions to this PR is to provide a command that can load a sub-command from a sub-command (e.g. the "foo" command extension could add its own "bar" command).